django/db/models/functions/comparison.py

```diff
@@ -175,21 +177,25 @@ def join(self, args):
         )
 
     def as_postgresql(self, compiler, connection, **extra_context):
-        if not connection.features.is_postgresql_16:
-            copy = self.copy()
+        copy = self.copy()
+        if (
+            not connection.features.is_postgresql_16
+            or connection.features.uses_server_side_binding
+        ):
             copy.set_source_expressions(
                 [
                     Cast(expression, TextField()) if index % 2 == 0 else expression
                     for index, expression in enumerate(copy.get_source_expressions())
                 ]
             )
-            return super(JSONObject, copy).as_sql(
-                compiler,
-                connection,
-                function="JSONB_BUILD_OBJECT",
-                **extra_context,
-            )
-        return self.as_native(compiler, connection, returning="JSONB", **extra_context)
+            if not connection.features.is_postgresql_16:
+                return super(JSONObject, copy).as_sql(
+                    compiler,
+                    connection,
+                    function="JSONB_BUILD_OBJECT",
+                    **extra_context,
+                )
+        return copy.as_native(compiler, connection, returning="JSONB", **extra_context)
```

Comment:
We don't need to create a copy to call `as_native()`. 
```suggestion
        return self.as_native(compiler, connection, returning="JSONB", **extra_context)
```